xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,165 @@
+
+        assert isa(
+            op.input_stencil.type,
+            memref.MemRefType[Attribute] | stencil.TempType[Attribute],
+        )
+        assert isa(
+            t_type := op.input_stencil.type.get_element_type(), TensorType[Attribute]
+        )
+
+        # when translating swaps, remove third dimension
+        prefetch_op = csl_stencil.PrefetchOp(
+            input_stencil=op.input_stencil.op,
+            size=IntegerAttr(uniform_size, IntegerType(64)),
+            topo=op.topo,
+            swaps=[
+                csl_stencil.ExchangeDeclarationAttr(swap.neighbor[:2])
+                for swap in op.swaps
+            ],
+            result_type=memref.MemRefType(
+                TensorType(t_type.get_element_type(), (uniform_size,)),
+                (len(op.swaps),),
+            ),
+        )
+
+        # a little hack to get around a check that prevents replacing a no-results op with an n-results op
+        rewriter.replace_matched_op(prefetch_op, new_results=[])
+
+        # uses have to be retrieved *before* the loop because of the rewriting happening inside the loop
+        uses = list(op.input_stencil.uses)
+
+        # csl_stencil.prefetch, unlike dmp.swap, has a return value. This is added as the last arg
+        # to stencil.apply, before rebuilding the op and replacing stencil.access ops by csl_stencil.access ops
+        # that reference the prefetched buffers (note, this is only done for neighbor accesses)
+        for use in uses:
+            if not isinstance(use.operation, stencil.ApplyOp):
+                continue
+            apply_op = use.operation
+
+            # arg_idx points to the stencil.temp type whose data is prefetched in a separate buffer
+            arg_idx = apply_op.args.index(op.input_stencil)
+
+            # add the prefetched buffer as the last arg to stencil.access
+            apply_op.region.block.insert_arg(
+                prefetch_op.result.type, len(apply_op.args)
+            )
+
+            # rebuild stencil.apply op
+            r_types = [r.type for r in apply_op.results]
+            assert isa(r_types, Sequence[stencil.TempType[Attribute]])
+            new_apply_op = stencil.ApplyOp.get(
+                list(apply_op.args) + [prefetch_op.result],
+                apply_op.region.clone(),
+                r_types,
+            )
+            rewriter.replace_op(apply_op, new_apply_op)
+
+            # replace stencil.access (operating on stencil.temp at arg_index)
+            # with csl_stencil.access (operating on memref at last arg index)
+            nested_rewriter = PatternRewriteWalker(
+                GreedyRewritePatternApplier(
+                    [
+                        AccessOpFromPrefetchPattern(arg_idx),
+                    ]
+                ),
+                walk_reverse=False,
+                apply_recursively=False,
+            )
+
+            nested_rewriter.rewrite_op(new_apply_op)
```

Comment:
I need to invoke it with that argument.

Reply:
Maybe this will become a simpler rewrite once we get the `csl_stencil.apply` ops?

It's usually a sign that the abstractions you are thinking with don't quite map to MLIRs cleanly. We can leave this in for now, but I think it's a sign that we are doing something not quite right yet.